network/discovery.go

```diff
@@ -218,52 +247,114 @@ func (d *discovery) getStream(peerID peer.ID) (interface{}, error) {
 	if err != nil {
 		return nil, err
 	}
 
-	addrInfo := make([]*peer.AddrInfo, len(resp.Nodes))
-
-	for indx, node := range resp.Nodes {
-		info, err := StringToAddrInfo(node)
-		if err != nil {
-			return nil, err
-		}
-
-		addrInfo[indx] = info
-	}
-
-	return addrInfo, nil
+	return resp.Nodes, nil
 }
 
 func (d *discovery) run() {
+	peerDiscoveryTicker := time.NewTicker(peerDiscoveryInterval)
+	bootnodeDiscoveryTicker := time.NewTicker(bootnodeDiscoveryInterval)
+
+	defer func() {
+		peerDiscoveryTicker.Stop()
+		bootnodeDiscoveryTicker.Stop()
+	}()
+
 	for {
 		select {
-		case <-time.After(5 * time.Second):
-		case <-d.notifyCh:
 		case <-d.closeCh:
 			return
+		case <-peerDiscoveryTicker.C:
+			go d.handleDiscovery()
+		case <-bootnodeDiscoveryTicker.C:
+			go d.bootnodeDiscovery()
 		}
-		d.handleDiscovery()
 	}
 }
 
 func (d *discovery) handleDiscovery() {
 	// take a random peer and find peers
-	if target := d.peers.getRandomPeer(); target != nil {
-		if err := d.attemptToFindPeers(target.id); err != nil {
-			d.srv.logger.Error("failed to dial peer", "peer", target.id, "err", err)
+	if d.srv.availableOutboundConns() > 0 {
+		if target := d.peers.getRandomPeer(); target != nil {
+			if err := d.attemptToFindPeers(target.id); err != nil {
+				d.srv.logger.Error("Failed to find new peers", "peer", target.id, "err", err)
+			}
+		}
+	}
+}
+
+// bootnodeDiscovery queries a random bootnode for new peers and adds them to the routing table
+func (d *discovery) bootnodeDiscovery() {
+	if d.srv.availableOutboundConns() <= 0 {
+		return
+	}
+
+	// get a random bootnode which is not connected
+	bootnode := d.srv.getBootNode()
+	if bootnode == nil {
+		return
+	}
+
+	// isTemporaryDial maintains the dial status
+	var isTemporaryDial bool
+
+	// if atleast one bootnode is connected the dial status is temporary
+	if d.bootnodeConnCount > 1 {
+		isTemporaryDial = true
+	}
+
+	if isTemporaryDial {
+		if _, loaded := d.srv.temporaryDials.LoadOrStore(bootnode.ID, true); loaded {
+			return
 		}
 	}
+
+	if len(d.srv.host.Peerstore().Addrs(bootnode.ID)) == 0 {
+		d.srv.host.Peerstore().AddAddr(bootnode.ID, bootnode.Addrs[0], peerstore.AddressTTL)
+	}
+
+	stream, err := d.srv.NewProtoStream(discProto, bootnode.ID)
+	if err != nil {
+		d.srv.logger.Error("Failed to open new stream", "peer", bootnode.ID, "err", err)
+
+		return
+	}
+
+	clt := proto.NewDiscoveryClient(stream.(*rawGrpc.ClientConn))
+
+	resp, err := clt.FindPeers(context.Background(), &proto.FindPeersReq{Count: defaultPeerReqCount})
+	if err != nil {
+		d.srv.logger.Error("Find peers call failed", "peer", bootnode.ID, "err", err)
+
+		return
+	}
+
+	if err := stream.(*rawGrpc.ClientConn).Close(); err != nil {
+		d.srv.logger.Error("Error closing grpc stream", "peer", bootnode.ID, "err", err)
+
+		return
+	}
+
+	if isTemporaryDial {
+		d.srv.temporaryDials.Delete(bootnode.ID)
+		d.srv.Disconnect(bootnode.ID, "Thank you")
+	}
```

Comment:
When I tried, ConnectedF was called after a node disconnected here. then node continued to connect to the peer.

Reply:
can you help me with the scenario.

Reply:
@munna0908 Please share the details of the scenario you have done during working on this PR.

Reply:
I tested with the following nodes:
node1: Bootnodes => node1, node2, node3
node2: Bootnodes => node1, node2
node3: Bootnodes => node3, node4
node4: Bootnodes => node3, node4

node1 won't be connecting to node3 first because node3 starts after node1 and node3 will not try to connect to node1. node1 will try to connect to node3 in bootnodeDiscovery after a while and query for its table. In this case, node1 was keeping connection after bootnodeDiscovery is done.

Reply:
@Kourin1996 I got the issue, fixed it ([link](https://github.com/0xPolygon/polygon-edge/pull/343/commits/677053d94aec7557bc7d7778900ad03280f4caf2))